test/e2e/capv_quick_start_test.go

```diff
@@ -37,3 +50,101 @@ var _ = Describe("Cluster Creation using Cluster API quick-start test", func() {
 		}
 	})
 })
+
+var _ = Describe("Cluster creation with vSphere validations", func() {
+	var namespace *corev1.Namespace
+
+	BeforeEach(func() {
+		Expect(bootstrapClusterProxy).NotTo(BeNil(), "BootstrapClusterProxy can't be nil")
+		namespace = setupSpecNamespace("capv-e2e")
+	})
+
+	AfterEach(func() {
+		cleanupSpecNamespace(namespace)
+	})
+
+	It("should create a cluster with storage policy", func() {
+		clusterName := fmt.Sprintf("cluster-%s", util.RandomString(6))
+		Expect(namespace).NotTo(BeNil())
+
+		By("creating a workload cluster")
+		configCluster := defaultConfigCluster(clusterName, namespace.Name)
+
+		_ = clusterctl.ApplyClusterTemplateAndWait(ctx, clusterctl.ApplyClusterTemplateAndWaitInput{
+			ClusterProxy:                 bootstrapClusterProxy,
+			ConfigCluster:                configCluster,
+			WaitForClusterIntervals:      e2eConfig.GetIntervals("", "wait-cluster"),
+			WaitForControlPlaneIntervals: e2eConfig.GetIntervals("", "wait-control-plane"),
+			WaitForMachineDeployments:    e2eConfig.GetIntervals("", "wait-worker-nodes"),
+		})
+
+		pbmClient, err := pbm.NewClient(ctx, vsphereClient.Client)
+		Expect(err).NotTo(HaveOccurred())
+		var res []types.PbmServerObjectRef
+		if pbmClient != nil {
+			spName := getVariable(VsphereStoragePolicy)
+			if spName == "" {
+				Fail("storage policy test run without setting VSPHERE_STORAGE_POLICY")
+			}
+
+			spID, err := pbmClient.ProfileIDByName(ctx, spName)
+			Expect(err).NotTo(HaveOccurred())
+
+			res, err = pbmClient.QueryAssociatedEntity(ctx, types.PbmProfileId{UniqueId: spID}, "virtualMachine")
+			Expect(err).NotTo(HaveOccurred())
+		}
+		Expect(len(res)).To(BeNumerically(">", 0))
+
+		vms := getVSphereVMsForCluster(clusterName, namespace.Name)
+		Expect(len(vms.Items)).To(BeNumerically(">", 0))
+
+		datacenter, err := vsphereFinder.DatacenterOrDefault(ctx, vsphereDatacenter)
+		Expect(err).ShouldNot(HaveOccurred())
+		By("verifying storage policy is used by VMs")
```

Comment:
the query for objects using the storage policy returns server UID along with object reference ID. The vsphereVM spec has the BIOS UID which is different so forced to fetch the vm again and compare object references. let me know if there's a simpler way I missed